### PR TITLE
feat(next, next-sample): support nextjs edge runtime

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add support for NextJS edge runtime (vercel):

1. remove lodash
2. bump iron-session to latest version
3. add `@logto/next/edge` sub package

## Testing

Local tested & deployed to vercel

## Checklist

- [ ] `.changset`
- [ ] unit tests
- [ ] integration tests
- [ ] docs

OR

- [x] This PR is not applicable for the checklist

### CODE REVIEW

File: packages/js/src/utils/errors.ts [hunk 1]
```diff
@@ -1,43 +1,26 @@
...
-    invalid_token: 'Invalid ID token',
-  },
-  callback_uri_verification: {
-    redirect_uri_mismatched: 'The callback URI mismatches the redirect URI.',
-    error_found: 'Error found in the callback URI',
-    missing_state: 'Missing state in the callback URI',
-    state_mismatched: 'State mismatched in the callback URI',
-    missing_code: 'Missing code in the callback URI',
-  },
+  id_token_invalid_iat: 'Invalid issued at time in the ID token',
```
Comment (gao-sun) [inline packages/js/src/utils/errors.ts:4]: i'm good to have breaking changes, as long as we have the major version bump

File: packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      return session;
+    },
+    set(value) {
+      const keys = Object.keys(value);
+      const currentKeys = Object.keys(session);
+
+      for (const key of currentKeys) {
+        if (!keys.includes(key)) {
+          /* eslint-disable @typescript-eslint/no-dynamic-delete */
+          /* eslint-disable @silverhand/fp/no-delete */
```
Comment (gao-sun) [inline packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts:18]:
```suggestion
          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete, @silverhand/fp/no-delete
```

File: packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          // @ts-expect-error See comment in IronSessionData interface
+          delete session[key];
+          /* eslint-enable @silverhand/fp/no-delete */
+          /* eslint-enable @typescript-eslint/no-dynamic-delete */
+        }
+      }
+
+      for (const key of keys) {
+        /* eslint-disable @typescript-eslint/no-unsafe-assignment */
+        /* eslint-disable @silverhand/fp/no-mutation */
```
Comment (gao-sun) [inline packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts:28]: ditto

File: packages/next/edge/iron-session-edge/index.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    // req.session.save();
+    // eslint-disable-next-line @silverhand/fp/no-mutating-methods
+    Object.defineProperty(request, 'session', getPropertyDescriptorForRequestSession(session));
+
+    return handler(request, response);
+  };
+}
+
+// Argument type based on the SSR context
+type GetIronSessionSSROptions = (
```
Comment (gao-sun) [inline packages/next/edge/iron-session-edge/index.ts:41]:
```suggestion
type GetIronSessionSsrOptions = (
```

File: packages/next/package.json [hunk 5]
```diff
@@ -3,14 +3,19 @@
...
-  },
   "module": "./lib/module.mjs",
   "types": "./lib/index.d.ts",
+  "exports": {
+    ".": {
+      "require": "./lib/index.js",
+      "import": "./lib/module.mjs",
+      "types": "./lib/index.d.ts"
+    },
+    "./edge": "./edge/lib/index.js"
```
Comment (gao-sun) [inline packages/next/package.json:14]: also "import" and "types"?

File: pnpm-lock.yaml [hunk 6]
```diff
@@ -9598,32 +9652,32 @@ packages:
       canvas:
         optional: true
     dependencies:
-      abab: 2.0.6
+      abab: 2.0.5
```
Comment (gao-sun) [inline pnpm-lock.yaml:9655]: saw a lot of version numbers are decreased, is it expected?

File: packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts (hunk 2 above)
Comment (wangsijie) [reply packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts:18]: But the next line is "@ts-expect-error", its effect is also in "next line"

File: pnpm-lock.yaml (hunk 6 above)
Comment (wangsijie) [reply pnpm-lock.yaml]: fixed

File: packages/next/package.json (hunk 5 above)
Comment (wangsijie) [reply packages/next/package.json:14]: If I add "import" and "types", parcel cannot generate them ("d.js" files), and will cause error for "import".

I can't figure out how to fix it, help needed!

File: packages/client/src/errors.ts [hunk 7]
```diff
@@ -1,34 +1,18 @@
-import type { NormalizeKeyPaths } from '@silverhand/essentials';
-import get from 'lodash.get';
-
 const logtoClientErrorCodes = Object.freeze({
-  sign_in_session: {
-    invalid: 'Invalid sign-in session.',
-    not_found: 'Sign-in session not found.',
-  },
+  sign_in_session_invalid: 'Invalid sign-in session.',
```
[the remark this answers is not in the document]
Comment (wangsijie) [reply packages/client/src/errors.ts]: I updated with a workaround to avoid a breaking change.

**Comment by wangsijie** [comment]
waiting for #473

**Comment by waltcow** [comment]
@wangsijie Hope it can be merged ASAP 😭

**Comment by wangsijie** [comment]
@waltcow there are some problems on the release flow, the good news is they are resolved, just need to rebase and resolve conflicts.

File: packages/next/package.json [hunk 8]
```diff
@@ -2,15 +2,24 @@
...
+      "import": "./lib/edge/index.mjs",
+      "types": "./lib/edge/index.d.ts"
+    }
   },
-  "module": "./lib/index.mjs",
-  "types": "./lib/index.d.ts",
   "files": [
-    "lib"
+    "lib",
+    "edge/lib"
```
Comment (gao-sun) [inline packages/next/package.json:22]: no need for this line?

File: packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts (hunk 2 above)
Comment (gao-sun) [reply packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts:18]: ah then you can switch the order i think:

```ts
// @ts-expect-error foo
// eslint-disable-next-line ...
```

**Review by gao-sun — APPROVED** [review]

File: packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts (hunk 2 above)
Comment (wangsijie) [reply packages/next/edge/iron-session-edge/get-property-descriptor-for-request-session.ts:18]: hhh, unfortunately, `@ts-expect-error` only affects the very next line

**Pull request merged**